### DISCUSSION TITLE
Parse texture from Material

### DISCUSSION THREAD

**Discussion by zrshannes** [donmccurdy/glTF-Transform#1554] [Q&A]
I have a material with texture, texCoord, and  intensity.
i wrote myself a extension for gltf-transform so "lightmap" Material will be read and writeable.

```
    {
      "doubleSided": true,
      "extensions": {
        "lightmap": {
          "intensity": 3.14,
          "index": 0,
          "texCoord": {
            "type": "Fixed",
            "value": "LM"
          }
        }
      },
      "name": "Material",
      "pbrMetallicRoughness": {
        "metallicFactor": 0,
        "roughnessFactor": 0.5
      }
    },
```

write:
```
        this.document.getRoot().listMaterials().forEach((material) => {
            if (material.getExtension(LIGHTMAP_EXTENSION_NAME)) {
                const materialIndex = context.materialIndexMap.get(material);
                const materialDef = jsonDoc.json.materials[materialIndex];
                materialDef.extensions = materialDef.extensions || {};

                const extensionDataLightmap = this.extensionDataLightmaps.get(materialIndex);
                if (extensionDataLightmap) {
                    console.log("\nWRITE Lightmap: \n", extensionDataLightmap, "\nto Scene Index", materialIndex);
                    materialDef.extensions[LIGHTMAP_EXTENSION_NAME] = extensionDataLightmap;
                    isLightmapUsed = true;
                }
            }
        });
```

read:
```
        const materialDefs = context.jsonDoc.json.materials || [];
        materialDefs.forEach((materialDef, index) => {
            if (materialDef.extensions && materialDef.extensions[LIGHTMAP_EXTENSION_NAME]) {
                this.extensionDataLightmaps.set(index, materialDef.extensions[LIGHTMAP_EXTENSION_NAME]);
                context.materials[index].setExtension(LIGHTMAP_EXTENSION_NAME, this.createLightmap());
            }
        });
```

when i process a glb the processed glb is missing the texture, buffers etc.
material is copied and image is also present and converted.
what am i missing to handle the given material correctly?

**Comment by donmccurdy** [comment]
Hi @zrshannes! I've converted this issue into a discussion. I would recommend implementing your extension by using one of the existing extensions as a reference. Another material extension that includes textures, like [`KHR_materials_anisotropy`](https://gltf-transform.dev/modules/extensions/classes/KHRMaterialsAnisotropy), would be a good starting point:

https://github.com/donmccurdy/glTF-Transform/tree/main/packages/extensions/src/khr-materials-anisotropy

I would note that the JSON structure you've included would be a bit non-standard, usually a goal of extension authors is that if the extension were to disappear (and all its properties added to the base material object) the names would still make sense. Instead, consider...

```jsonc
{
    "materials": [
        {
            "extensions": {
                "MYPREFIX_materials_lightmap": {
                    "lightmapIntensity": 3.14,
                    "lightmapTexture": { "index": 0 }
                }
            }
        }
    ]
}
```

... which I think you may find easier, as the `lightmapTexture` entry maps to an existing [`TextureInfo`](https://gltf-transform.dev/modules/core/classes/TextureInfo) class.

**Reply by zrshannes** [reply]
sadly the json structure is set in stone.
i also came across anisotropy as reference.

i wrap my head around it and see if i can figure something out.
but really, thanks for your reply!
